Matching-Collab Integration & Edge Case Handling

CJianzhi opened the pull request:
# Documentation
I have wrote the full flow documentation for those that want to understand what happens from matching all the way to the code editor stage behind the scenes in the shared google drive. Detailed implementation of the edge case is also written there. However, I am not sure if our current approach discussed during Tuesday's meeting will lead to any penalisation from the teaching team due to the quite extensive coupling.

https://docs.google.com/document/d/1AtB0mZW_TT_Jnr_llNLLuMsTs1sYFDxRUgyHpjRX1jg/edit?tab=t.0

# Known issue:
There is some lag before the user navigates to collab page even after the creating room dialog box closes. This is likely due to the turbopack rebuilding which only occurs in development which I'm not sure why it happens.(based on GPT) 

# Testing
Do copy the env.sample into the env files for each service (except user service get from drive as it contains hidden secret information)

# List of Changes

## Major changes
1) Add redis for persistence storage of session data to accmodate server crashes
2) Matching service calls collab service to create room first before socket connection 
3) Shift socket connection to matching page instead of CodeComponent and create a WebSocketManager singleton class to manage websocket connection
4) Handle disconnection and rejoining room and other edge cases listed below 
5) Add JWT authentication for socket connection and collab service endpoints
6) Currently, the only way to access the collab page is by matching with another user OR rejoining the room when prompted after accidental disconnection

## Minor changes
1) Add buffer time of 2 minutes after both useres leave session before deleting session data in Redis and local memory
2) Resolve socket connection for cloud deployment 5 minute timeout issue (can only be tested by deploying)
3) Alert shows up to inform user that partner left session
4) Add confirmation dialog box when leave session button is clicked

fixes #52 , #44 , #40 


# Edge Cases 

The google drive docs contains explanation for what each useEffect does

## Case 1: Leaving session intentionally thru leave button

Behaviour: Ask for confirmation, on confirmation, redirect to matching page

## Case 2: Leaving session by reloading

Behaviour: When user reloads the page, ~~they are brought back to matching page and pop up appears asking if user wants to rejoin the session.~~  a black screen is displayed with an "unauthorised access" dialog box and user is navigated back to the matching page if they click back.

## Case 3: Leaving session by closing browser

Behaviour: When they go back to matching page, pop up appears asking if user wants to rejoin the session. A session can only be rejoined if the partner has not left. 

## Case 4: When user tries to access collab page after not wanting to rejoin a session, or after leaving intentionally by clicking the leave session button, or by trying to change the browser URL

Behaviour: A black screen is displayed with an "unauthorised access" dialog box and user is navigated back to the matching page if they click back.

## Case 5: Server crash (partial implemetation)

Behaviour: When server crash, sessionData stored on local memory is deleted, hence on server start up, all sessionData from redis will be loaded into the local memory map. However, I have yet to implement what happens on the frontend when server crash and how user can recover

## Case 6: Multiple tabs opened by same user (Not handled yet)

## Case 7: wifi disconnected / offline editing (Not handled yet)

FooChao commented on .env.sample:
```diff
@@ -1,9 +0,0 @@
-REDIS_PASSWORD=example
```
please dun delete .env.sample it is to indicate that .env is expected

FooChao reviewed (commented):
lgtm. nice work. i checked the code logic it is clear and good.

It would be great if u can inform others to update the .env in the pr description.

Just some UI/UX suggestion which u can ignore as they are not super important:

- I find it sudden that when on wrong collab url it is redirected straight to matching with no information given. it will be great if u can give dialog or direct to an error page instead but based on my understanding of code it can be a bit tedious to do it.

- when teammate disconnect and then connect back, user was not informed about it.

If possible can u change the hard coded url back to the original dynamic one else it will crash my deployment

CJianzhi commented on matching-service/.env.sample:
```diff
@@ -1,6 +1,8 @@
 PORT=3002
 REDIS_HOST=redis://redis-matching-service:6379
 
+COLLAB_SERVICE_URL=http://collab-service:8000/api
```
Yes it is intentional because I thought the gateway is mainly for client - service communication and not used for inter service communication. I think removing the /api is a good idea, thanks for the suggestion

https://stackoverflow.com/questions/53545963/internal-communication-microservices-through-api-gateway

LemonDrew reviewed (dismissed):
LGTM! 

The google doc that you provided helped me tremendously in tracing through the code. Speaking of which, we should probably add the sequence diagram for each service as well.

Do note that the env sample for user service doesnt work if pasted into .env directly

Transition from matching to collab page is working well. To be honest, I kinda like the dialog stating that the cursor and room is setting up as it tells the user when the room is ready and the duration at which the dialog is open is negligible.  

Tested all 3 use cases (Intentionally leaving, refresh and closing browser) and all are working well.

CJianzhi commented:
Made some changes based on review comments and added new commits

1) Added back .env.sample for root directory
2) Modified api gateway default.config file to use dynamic variable instead of hard coded front end URL
3) Added an unauthorised dialog box where user has to click "back" button in order to navigate back to matching page when they try to access collab page so that the navigation is not sudden. (implementation is a little tricky and details added to the docs because I'm not sure if the current way to implement this is the best way, but its one of the ways I can think of)
4) Fixed bug identified by @ITZXITZX where if user chooses to leave after being notified that partner has left, the "creating room" dialog box appears when they start a new matching process

I will address other edge cases mentioned above in another PR so that this PR can be closed once bug free and approved.
For your perusal and approval  @FooChao @LemonDrew

FooChao reviewed (dismissed):
lgtm!
changed the redis set up to be like matching so it can be deployed
I have updated the gcp env for matching and collab so it should work after merge (I hope)
collab and matching env changed again @LemonDrew just to let u know.
I spent forever figuring out why it keep failing
